autogen/oai/gemini.py

```diff
@@ -261,29 +268,67 @@ def create(self, params: Dict) -> ChatCompletion:
                 completion_tokens=completion_tokens,
                 total_tokens=prompt_tokens + completion_tokens,
             ),
-            cost=calculate_gemini_cost(prompt_tokens, completion_tokens, model_name),
+            cost=self._calculate_gemini_cost(prompt_tokens, completion_tokens, model_name),
         )
 
         return response_oai
-
-    def _oai_content_to_gemini_content(self, content: Union[str, List]) -> List:
+    
+    def _oai_content_to_gemini_content(self, message: Dict[str, Any]) -> List:
         """Convert content from OAI format to Gemini format"""
         rst = []
-        if isinstance(content, str):
+        if isinstance(message, str):
+            if self.use_vertexai:
+                rst.append(VertexAIPart.from_text(message))
+            else:
+                rst.append(Part(text=message))
+            return rst
+
+        if "tool_calls" in message:
+            if self.use_vertexai:
+                rst.append(VertexAIPart.from_dict({
+                    "functionCall": {
+                        "name": message["tool_calls"][0]["function"]["name"],
+                        "args": json.loads(message["tool_calls"][0]["function"]["arguments"])
+                    }
+                }))
+            else:
+                rst.append(
+                    Part(
+                        function_call=FunctionCall(
+                            name=message["tool_calls"][0]["function"]["name"],
+                            args=json.loads(message["tool_calls"][0]["function"]["arguments"]),
+                        )
+                    )
+                )
+            return rst
+
+        if message["role"] == "tool":
             if self.use_vertexai:
-                rst.append(VertexAIPart.from_text(content))
+                rst.append(VertexAIPart.from_function_response(
+                    name=message["name"],
+                    response=json.loads(message["content"])
```

Comment:
@arjun-g note that because of the way Gemini from_function_response expects input we're going to have to either change the [example calculator function](https://microsoft.github.io/autogen/docs/tutorial/tool-use/#creating-tools) or do something like the following in this file:
```
# Prepare the response in the expected dictionary format
response = {"content": {"answer": str(result)}}
```

Otherwise folks who try to run the calculator example with Gemini will get the following error:
```
  File "/Users/drewgillson/autogen-gemini/venv/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 2032, in from_function_response
    function_response=gapic_tool_types.FunctionResponse(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/drewgillson/autogen-gemini/venv/lib/python3.12/site-packages/proto/message.py", line 581, in __init__
    pb_value = marshal.to_proto(pb_type, value)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/drewgillson/autogen-gemini/venv/lib/python3.12/site-packages/proto/marshal/marshal.py", line 228, in to_proto
    pb_value = self.get_rule(proto_type=proto_type).to_proto(value)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/drewgillson/autogen-gemini/venv/lib/python3.12/site-packages/proto/marshal/rules/struct.py", line 140, in to_proto
    k: self._marshal.to_proto(struct_pb2.Value, v) for k, v in value.items()
                                                               ^^^^^^^^^^^
AttributeError: 'float' object has no attribute 'items'
```

Reply:
Fixed this. Changed the function response to a dict in this format

`{"result":  #result}`